manual/manual/html_processing/src/process_manual.ml

```diff
@@ -0,0 +1,465 @@
+  match Re.Str.search_forward charset_regexp html 0 with
+  | exception Not_found -> dbg "Warning, no charset found in html."; html
+  | _ -> match (String.lowercase_ascii (Re.Str.matched_group 1 html)) with
+    | "utf-8" -> dbg "Charset is UTF-8; good."; html
+    | "us-ascii" -> dbg "Charset is US-ASCII. We change it to UTF-8";
+        Re.Str.global_replace charset_regexp "charset=UTF-8\\2" html
+    | _ -> dbg "Warning, charset not recognized."; html
+
+(* Save new html file *)
+let save_to_file soup file =
+  let new_html = to_string soup in
+  write_file (docs_file file) new_html
+
+(* Find title associated with file *)
+let file_title file toc =
+  if file = "index.html" then Some index_title
+  else Hashtbl.find_opt toc file
+
+(* Replace three links "Previous, Up, Next" at the end of the file by more
+   useful titles, and insert then in a div container, keeping only 2 of them:
+   either (previous, next) or (previous, up) or (up, next). Remove them at the
+   top of the file, where they are not needed because we have the TOC. *)
+let update_navigation soup toc =
+  Option.iter delete (soup $? "hr");
+  let container, count =
+    ["Previous"; "Up"; "Next"]
+    |> List.fold_left (fun (container, count) s ->
+        let imgs = soup $$ ("img[alt=\"" ^ s ^ "\"]") in
+        (* In principle [imgs] will contain either 0 or 2 elements. We delete
+           the first one. *)
+        Option.iter (delete << R.parent) (first imgs);
+        (* Now, if there is a second element, we update (or create) an "div"
+           container to insert the element, and increase [count]. [count] is
+           used to make sure we keep only 2 links, not 3. *)
+        imgs |> fold (fun (container, count) img ->
+            let a = R.parent img in
+            let file = R.attribute "href" a in
+            let title = match file_title file toc with
+              | Some f -> begin match s with
+                  | "Previous" -> "« " ^ f
+                  | "Next" -> f ^ " »"
+                  | "Up" -> f
+                  | _ -> failwith "This should not happen"
+                end
+              | None -> dbg "Unknown title for file %s" file; s in
+            let txt = create_text title in
+            add_class (String.lowercase_ascii s) a;
+            replace img txt;
+            let div = match container with
+              | Some c ->
+                  if count = 2 then delete (children c |> R.last);
+                  (* : we delete the "Up" link *)
+                  append_child c a; c
+              | None ->
+                  let c = create_element ~class_:"bottom-navigation" "div" in
+                  wrap a c; c in
+            Some div, count+1) (container, count)) (None, 0) in
+
+  match container with
+  | None -> dbg "%s" "No navigation"
+  | Some div ->
+      if count = 2 then begin
+        add_class "previous" (div $$ "a" |> R.first);
+        add_class "next" (div $$ "a" |> R.last);
+      end
```

Comment:
The double fold seems too complicated. You could replace it by:

* fetch all previous, next, up, images
* filter the case 0 or 2 elements
* eliminate the version of the images at the top.
* Select 2 elements among the non-empty elements of the triplets
* Replace the first non-empty element with the new nav items.

Reply:
ok, I've simplified this along your lines. It's not shorter, but probably less convoluted indeed.